overlord/snapstate/conflict.go

```diff
@@ -0,0 +1,161 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// ChangeConflictError represents an error because of snap conflicts between changes.
+type ChangeConflictError struct {
+	Snap       string
+	ChangeKind string
+	// a Message is optional, otherwise one is composed from the other information
+	Message string
+}
+
+func (e *ChangeConflictError) Error() string {
+	if e.Message != "" {
+		return e.Message
+	}
+	if e.ChangeKind != "" {
+		return fmt.Sprintf("snap %q has %q change in progress", e.Snap, e.ChangeKind)
+	}
+	return fmt.Sprintf("snap %q has changes in progress", e.Snap)
+}
+
+// An AffectedSnapsFunc returns a list of affected snap names for the given supported task.
+type AffectedSnapsFunc func(*state.Task) ([]string, error)
+
+var (
+	affectedSnapsByAttr = make(map[string]AffectedSnapsFunc)
+	affectedSnapsByKind = make(map[string]AffectedSnapsFunc)
+)
+
+// AddAffectedSnapsByAttrs registers an AffectedSnapsFunc for returning the affected snaps for tasks sporting the given identifying attribute, to use in conflicts detection.
+func AddAffectedSnapsByAttr(attr string, f AffectedSnapsFunc) {
+	affectedSnapsByAttr[attr] = f
```

Comment:
Since this is public and any subsystem can set own functions, I think we should make sure a mapping here is not accidently overwritten by a programming mistake. IMHO a panic would be sufficient if we see an attempt to set function for already existing attr.

Reply:
that is annoying for tests where this is called again with the same values, for example by the hookstate code

Reply:
Fair enough, disregard; pain factor vs benefit is not worth it, plus we've similiar approach when it comes to task runners for example.